context_test.go

```diff
@@ -180,6 +180,24 @@ func TestContextBuilderParentContextReplaceHeaders(t *testing.T) {
 		SetHeaders(map[string]string{"header key": "header value 2"}).
 		Build()
 	assert.Equal(t, map[string]string{"header key": "header value 2"}, ctx2.Headers())
+
+	goroutines.VerifyNoLeaks(t, nil)
+}
+
+func TestContextWrapWithHeaders(t *testing.T) {
+	headers1 := map[string]string{
+		"k1": "v1",
+	}
+	ctx, _ := NewContextBuilder(time.Second).
+		SetHeaders(headers1).
+		Build()
+	assert.Equal(t, headers1, ctx.Headers(), "Headers mismatch after Build")
+
+	headers2 := map[string]string{
+		"k1": "v1",
+	}
+	ctx2 := WrapWithHeaders(ctx, headers2)
+	assert.Equal(t, headers2, ctx2.Headers(), "Headers mismatch after WrapWithHeaders")
```

Comment:
This one doesn't cause `context.WithCancel` or `context.WithTimeout` to be called, so it can't leak a goroutine. The unwrapping in this method is just to avoid having a structure like:
```
context.Context -> headerCtx -> headerCtx -> headerCtx
```
Instead we want to flatten to
```
context.Context -> headerCtx
```

(`headerCtx` should always be the last context and we should only need one)